llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -4933,11 +4933,17 @@ bool LoopVectorizationPlanner::isMoreProfitable(
       EstimatedWidthB *= *VScale;
   }
 
+  if (MaxTripCount > 0) {
+    EstimatedWidthA = std::min(EstimatedWidthA, MaxTripCount);
```

Comment:
`LoopVectorizationCostModel::getMaximizedVFForTarget` should limit the maximum VFs so that no VF exceeds at the loop's trip count. Would it be possible to handle minimum VSCale there instead?

There's already accounts for the `vscale` function attribute there

```
  unsigned WidestRegisterMinEC = MaxVectorElementCount.getKnownMinValue();
  if (MaxVectorElementCount.isScalable() &&
      TheFunction->hasFnAttribute(Attribute::VScaleRange)) {
    auto Attr = TheFunction->getFnAttribute(Attribute::VScaleRange);
    auto Min = Attr.getVScaleRangeMin();
    WidestRegisterMinEC *= Min;
  }
```

also cc @david-arm

Reply:
Ah yes, the cases I was thinking about were limiting to a smaller *scalable* VF.